src/main/scala/cognite/spark/v1/MetricsSource.scala

```diff
@@ -2,22 +2,36 @@ package org.apache.spark.datasource
 
 import cats.Eval
 import com.codahale.metrics._
-import org.apache.spark._
-
 import java.util.concurrent.ConcurrentHashMap
+import org.apache.spark._
 
 class MetricsSource {
   // Add metricNamespace to differentiate with spark system metrics.
   // Keeps track of all the Metric instances that are being published
   val metricsMap = new ConcurrentHashMap[String, Eval[Counter]]
 
-  def getOrCreateCounter(metricNamespace: String, metricName: String): Counter = {
+  def getOrCreateCounter(metricNamespace: String, name: String): Counter = {
+    val ctx = Option(TaskContext.get())
+
+    def getNumberOrEmpty(getter: TaskContext => AnyVal): String =
+      ctx.map(getter).map(_.toString()).getOrElse("")
+
+    // The number of fields in the name should be consistent, even if there is no value for
+    // the attempt numbers
+    val stageId = getNumberOrEmpty(_.stageId())
+    val stageAttempt = getNumberOrEmpty(_.stageAttemptNumber())
+    val partitionId = getNumberOrEmpty(_.partitionId())
+    val taskAttempt = getNumberOrEmpty(_.taskAttemptId())
+
+    val metricName = s"<$stageId>[$stageAttempt]#<$partitionId>[$taskAttempt]#$name"
```

Comment:
for a big format like that it might be nice to put the reading part also in this lib + a test or two to illustrate the format (normal case, case with empty strings). Also maybe since fields are so many a small data type to represent taskContext+name

Only a little surprised that current tests didn't notice much of a change. But to hypothesize, if a consumer of old metric format sees it it will interpret the whole 5-part thing as "metricName"? Might be especially interesting how they interact in case metrics are persisted and either old or new reader might see both kinds in transition periods.

Also had a look at https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md - exposition format probably doesn't matter a lot as it should encode, but https://prometheus.io/docs/concepts/data_model/ metric name restrictions might be important as not sure what will happen when disallowed prometheus-metric-name characters are used (would they be replaced, converted, or will something fail?). Not seeing an explicit length restriction there. Would feel more safe if we stay within allowed prometheus names, so maybe some variation on `::`? Like `":stage_:$stageId:stagetry_:$stageAttempt:part_:$partitionId:tasktry_:$taskAttempt:$name" ?`

Reply:
Good point. Will change to not risk having the name change from underneath our feet.
As for the tests, the brunt of it happened it the repo that uses this as a datasource

Will also bring the parser here.